File: tools/convsym/src/output/DEB2.cpp

/* ------------------------------------------------------------ *
 * ConvSym utility version 2.12									*
 * Output wrapper for the Debug Information format version 2.0	*
 * ------------------------------------------------------------	*/

#include <cassert>
#include <map>
#include <cstdint>
#include <string>
#include <vector>

#include <OptsParser.hpp>
#include <Huffman.hpp>
#include <BitStream.hpp>
#include <IO.hpp>
#include <utils.hpp>

#include "OutputWrapper.hpp"


struct Output__Deb2 : public OutputWrapper {

	Output__Deb2() {};
	~Output__Deb2() {};

	/**
	 * Main function that generates the output
	 */
	void parse(
		std::multimap<uint32_t, std::string>& symbols,
		const char * fileName,
		uint32_t appendOffset = 0,
		uint32_t pointerOffset = 0,
		const char * opts = "",
		bool alignOnAppend = true
	) {
		assert(!symbols.empty());

		auto output = OutputWrapper::setupOutput( fileName, appendOffset, pointerOffset, alignOnAppend );

		/* Parse options from "-inopt" agrument's value */
		bool optFavorLastLabels = false;

		const std::map<std::string, OptsParser::record>
			OptsList {
				{ "favorLastLabels",	{ .type = OptsParser::record::p_bool,	.target = &optFavorLastLabels	} }
			};
		
		OptsParser::parse( opts, OptsList );

		/* Write format version token */
		output->writeBEWord(0xDEB2);

		/* Allocate space for blocks offsets table */
		auto lastSymbolPtr = symbols.rbegin();
		uint16_t lastBlock = (lastSymbolPtr->first) >> 16;

		if (lastBlock > 0xFF) {		// blocks index table is limited to $100 entries (which is enough to cover all the 24-bit addressable space)
			IO::Log(IO::error, "Too many memory blocks to allocate ($%X), truncating to $100 blocks. Some symbols will be lost.", lastBlock+1);
			lastBlock = 0xFF;
		}

		std::vector<uint32_t> blockOffsets(lastBlock+1);
		const size_t blockOffsetsSize = blockOffsets.size() * sizeof(uint32_t);

		output->writeBEWord( blockOffsetsSize + 2 );				// write relative pointer to the Huffman table
		uint32_t loc_BlockOffsets = output->getCurrentOffset();	// remember the offset where blocks offset table should start
		output->setOffset( blockOffsetsSize, IO::current );		// reserve space to write down offsets table later

		/* ------------------------------------------------ */
		/* Generate Huffman-codes and create decoding table */
		/* ------------------------------------------------ */

		IO::Log( IO::debug, "Building an encoding table...");

		/* Generate table of character frequencies based on symbol names */
		uint32_t freqTable[0x100] = { 0 };
		for (auto& symbol : symbols) {
			for (auto& character : symbol.second) {
				freqTable[(int)character]++;
			}
			freqTable[0x00]++;	// include null-terminator
			// TODOh: Guess whether NULL will be appended to the string of specified length
		}

		/* Generate table of Huffman codes (sorted by code) */
		Huffman::RecordSet codesTable = Huffman::encode( freqTable );

		/* Write down the decoding table */
		const Huffman::Record* characterToRecord[0x100] = { nullptr };	// LUT that links each character to its Huffman-coding record
		for (auto& entry : codesTable) {
			if (entry.codeLength > 16) {
				const char * errorMessage = "Some encoding table code lengths exceed 16 bits, try -tolower or -toupper option to reduce entropy";
				throw errorMessage;
			}

			characterToRecord[entry.data] = &entry;		// assign character this Huffman::Record entity
			output->writeBEWord(entry.code);			// write Huffman-code
			output->writeByte(entry.codeLength);		// write Huffman-code length (in bits)
			output->writeByte(entry.data);				// write down original character
		}
		output->writeWord(-1);			// write 0xFFFF at the end of Huffman-table to stop searching and cause error

		/* ------------------------------------- */
		/* Generate per block symbol information */
		/* ------------------------------------- */

		{
			IO::Log(IO::debug, "Generating symbol data blocks...");

			auto symbolPtr = symbols.begin();
			struct SymbolRecord { uint16_t offset; uint16_t symbolDataPtr; };

			/* For 64kb block within symbols range */
			for (uint16_t block = 0x00; block <= lastBlock; block++) { 
				/* Align block on even address */
				if (output->getCurrentOffset() & 1) {
					output->writeByte(0x00);
				}

				uint32_t loc_Block = output->getCurrentOffset();	// remember offset, where this block starts ...

				BitStream SymbolsHeap;
				std::vector<SymbolRecord> offsetsData;

				/* For every symbol within the block ... */
				for (; (symbolPtr->first>>16) <= block && (symbolPtr != symbols.cend()); ++symbolPtr) {
					if ((symbolPtr->first>>16) < block) {
						continue;
					}

					/* 
					 * For records with the same offsets, fetch only the last or the first processed symbol,
					 * depending "favor last labels" option ...
					 */
					if ((optFavorLastLabels && std::next(symbolPtr) != symbols.end()
							&& std::next(symbolPtr)->first == symbolPtr->first) ||
						 (!optFavorLastLabels && symbolPtr != symbols.begin()
							&& std::prev(symbolPtr)->first == symbolPtr->first)
					) {
						continue;
					}

					if (SymbolsHeap.getCurrentPos() > 0xFFFF) {
						IO::Log(IO::error,"Symbols heap for block %02X exceeded 64kb limit, no more symbols can be stored in this block.", block);
						break;
					}
					
					else if (offsetsData.size() > 0x3FFF) {
						IO::Log(IO::error, "Too many symbols in block %02X, no more symbols can be stored in this block.", block);
						break;
					}

					else {
						/* Generate symbol structure, that includes offset and encoded symbol text pointer */
						offsetsData.push_back({
							.offset = (uint16_t)(symbolPtr->first & 0xFFFF),
							.symbolDataPtr = (uint16_t)SymbolsHeap.getCurrentPos()
						});
	
						/* Encode each symbol character with the generated Huffman-codes and store it in the bitsteam */
						for (auto& character : symbolPtr->second) {
							auto *record = characterToRecord[ (int)character ];
							SymbolsHeap.pushCode( record->code, record->codeLength );
						}
						
						/* Finally, add null-terminator */
						{
							auto *record = characterToRecord[ 0x00 ];
							SymbolsHeap.pushCode( record->code, record->codeLength );
							SymbolsHeap.flush();
						}
					}

				}

				/* Write offsets block and their corresponding encoded symbols heap */
				if ( offsetsData.size() > 0 ) {
					IO::Log(IO::debug,
						"Block %02X: %d bytes (offsets list: %d bytes, symbols heap: %d bytes)",
						block, 2+offsetsData.size()*4+SymbolsHeap.size(), offsetsData.size()*4, SymbolsHeap.size()
					);

					/* Insert pointer to the end of the list (where heap starts) */
					output->writeBEWord(2 + offsetsData.size()*4);		// write down offset

					/* Write down records that store offset and symbol pointer */
					for (auto& entry : offsetsData) {
						output->writeBEWord(entry.offset);			// write down offset
						output->writeBEWord(entry.symbolDataPtr);	// write down expected relative pointer to the heap
					}

					output->writeData(SymbolsHeap.begin(), SymbolsHeap.size());
					blockOffsets[block] = swap32((loc_Block-loc_BlockOffsets));
				}
			}
		}
		
		/* Finally, write down block offsets table in the header */
		output->setOffset(loc_BlockOffsets, IO::start);
		output->writeData(blockOffsets.data(), blockOffsetsSize);
	};

};

File: tools/crates/convsym/src/output/deb2.rs
/// Output generator for Debug Information format version 2.0 (DEB2).
/// Uses Huffman encoding for symbol names, improved over DEB1.

use std::collections::HashMap;
use std::io::{Seek, SeekFrom, Write};

use anyhow::Result;

trait WriteSeek: Write + Seek {}
impl<T: Write + Seek> WriteSeek for T {}

use crate::bitstream::BitStream;
use crate::huffman;
use crate::opts_parser::{self, RecordTarget};
use crate::output::OutputGenerator;

pub struct Deb2Output;

fn write_be_word(w: &mut impl Write, val: u16) -> Result<()> {
    w.write_all(&val.to_be_bytes())?;
    Ok(())
}

fn write_word(w: &mut impl Write, val: u16) -> Result<()> {
    w.write_all(&val.to_le_bytes())?;
    Ok(())
}

fn swap32(x: u32) -> u32 {
    (x >> 24) | ((x >> 8) & 0xFF00) | ((x << 8) & 0xFF0000) | (x << 24)
}

impl OutputGenerator for Deb2Output {
    fn generate(
        &self,
        symbols: &[(u32, String)],
        file_name: &str,
        append_offset: i64,
        pointer_offset: u32,
        opts: &str,
        align_on_append: bool,
    ) -> Result<()> {
        if symbols.is_empty() {
            anyhow::bail!("No symbols to output");
        }

        let mut output = setup_output(file_name, append_offset, pointer_offset, align_on_append)?;

        // Options
        let mut favor_last_labels = false;
        {
            let mut records: HashMap<&str, RecordTarget> = HashMap::new();
            records.insert(
                "favorLastLabels",
                RecordTarget::Bool(&mut favor_last_labels),
            );
            opts_parser::parse(opts, &mut records);
        }

        // Write format version token
        write_be_word(&mut output.writer, 0xDEB2)?;
        output.pos += 2;

        // Calculate last block
        let last_symbol_offset = symbols.last().unwrap().0;
        let mut last_block = (last_symbol_offset >> 16) as u16;
        if last_block > 0xFF {
            eprintln!(
                "Error: Too many memory blocks (${:X}), truncating to $100 blocks",
                last_block + 1
            );
            last_block = 0xFF;
        }

        let block_count = last_block as usize + 1;
        let block_offsets_size = block_count * 4; // 4 bytes per block offset

        // Write relative pointer to the Huffman table
        write_be_word(&mut output.writer, (block_offsets_size + 2) as u16)?;
        output.pos += 2;

        let loc_block_offsets = output.pos;

        // Reserve space for block offsets
        output.writer.write_all(&vec![0u8; block_offsets_size])?;
        output.pos += block_offsets_size;

        // Build frequency table
        let mut freq_table = [0u32; 256];
        for (_, label) in symbols {
            for &b in label.as_bytes() {
                freq_table[b as usize] += 1;
            }
            freq_table[0] += 1; // null terminator
        }

        // Generate Huffman codes
        let codes_table = huffman::encode(&freq_table, 16);

        // Build character-to-record LUT
        let mut char_to_record: [Option<(u32, u8)>; 256] = [None; 256];
        for record in &codes_table {
            if record.code_length > 16 {
                anyhow::bail!(
                    "Some encoding table code lengths exceed 16 bits, try -tolower or -toupper option to reduce entropy"
                );
            }
            char_to_record[record.data as usize] = Some((record.code, record.code_length));

            // Write Huffman table entry
            write_be_word(&mut output.writer, record.code as u16)?;
            output.writer.write_all(&[record.code_length])?;
            output.writer.write_all(&[record.data as u8])?;
            output.pos += 4;
        }
        // Write 0xFFFF terminator
        write_word(&mut output.writer, 0xFFFF)?;
        output.pos += 2;

        // Generate per-block symbol information
        let mut sym_idx = 0;
        let mut block_offsets_data: Vec<u32> = vec![0u32; block_count];

        for block in 0..=last_block {
            // Align on even address
            if output.pos & 1 != 0 {
                output.writer.write_all(&[0])?;
                output.pos += 1;
            }

            let loc_block = output.pos;

            let mut symbols_heap = BitStream::new();
            struct SymbolRecord {
                offset: u16,
                symbol_data_ptr: u16,
            }
            let mut offsets_data: Vec<SymbolRecord> = Vec::new();

            while sym_idx < symbols.len() {
                let (offset, ref label) = symbols[sym_idx];
                let sym_block = (offset >> 16) as u16;

                if sym_block > block {
                    break;
                }
                if sym_block < block {
                    sym_idx += 1;
                    continue;
                }

                // Handle duplicate offsets
                if should_skip_duplicate(symbols, sym_idx, favor_last_labels) {
                    sym_idx += 1;
                    continue;
                }

                if symbols_heap.get_current_pos() > 0xFFFF {
                    eprintln!(
                        "Error: Symbols heap for block {:02X} exceeded 64kb limit",
                        block
                    );
                    break;
                }

                if offsets_data.len() > 0x3FFF {
                    eprintln!(
                        "Error: Too many symbols in block {:02X}",
                        block
                    );
                    break;
                }

                offsets_data.push(SymbolRecord {
                    offset: (offset & 0xFFFF) as u16,
                    symbol_data_ptr: symbols_heap.get_current_pos() as u16,
                });

                // Encode symbol characters
                for &b in label.as_bytes() {
                    if let Some((code, code_len)) = char_to_record[b as usize] {
                        symbols_heap.push_code(code, code_len);
                    }
                }
                // Null terminator
                if let Some((code, code_len)) = char_to_record[0] {
                    symbols_heap.push_code(code, code_len);
                    symbols_heap.flush();
                }

                sym_idx += 1;
            }

            if !offsets_data.is_empty() {
                // Write list header: pointer to end of list (where heap starts)
                let list_end_offset = 2 + offsets_data.len() as u16 * 4;
                write_be_word(&mut output.writer, list_end_offset)?;
                output.pos += 2;

                // Write offset/pointer records
                for entry in &offsets_data {
                    write_be_word(&mut output.writer, entry.offset)?;
                    write_be_word(&mut output.writer, entry.symbol_data_ptr)?;
                    output.pos += 4;
                }

                // Write symbols heap data
                output.writer.write_all(symbols_heap.data())?;
                output.pos += symbols_heap.size();

                block_offsets_data[block as usize] =
                    swap32((loc_block - loc_block_offsets) as u32);
            }
        }

        // Write block offsets table at reserved position
        output.writer.seek(SeekFrom::Start(
            (output.base_offset + loc_block_offsets) as u64,
        ))?;
        for off in &block_offsets_data {
            output.writer.write_all(&off.to_ne_bytes())?;
        }

        Ok(())
    }
}

fn should_skip_duplicate(symbols: &[(u32, String)], idx: usize, favor_last: bool) -> bool {
    if favor_last {
        if idx + 1 < symbols.len() && symbols[idx + 1].0 == symbols[idx].0 {
            return true;
        }
    } else {
        if idx > 0 && symbols[idx - 1].0 == symbols[idx].0 {
            return true;
        }
    }
    false
}

struct OutputFile {
    writer: Box<dyn WriteSeek>,
    base_offset: usize,
    pos: usize,
}

fn setup_output(
    file_name: &str,
    append_offset: i64,
    pointer_offset: u32,
    align_on_append: bool,
) -> Result<OutputFile> {
    if append_offset != 0 {
        let mut file = std::fs::OpenOptions::new()
            .read(true)
            .write(true)
            .open(file_name)?;

        let actual_append;
        if append_offset == -1 {
            let end = file.seek(SeekFrom::End(0))? as usize;
            actual_append = if align_on_append && (end & 1) != 0 {
                file.write_all(&[0])?;
                end + 1
            } else {
                end
            };
        } else {
            actual_append = append_offset as usize;
            if align_on_append && (actual_append & 1) != 0 {
                eprintln!("Warning: An odd append offset is specified; the offset wasn't auto-aligned.");
            }
            file.seek(SeekFrom::Start(actual_append as u64))?;
        }

        if pointer_offset != 0 {
            file.seek(SeekFrom::Start(pointer_offset as u64))?;
            file.write_all(&(actual_append as u32).to_be_bytes())?;
        }

        file.seek(SeekFrom::Start(actual_append as u64))?;

        Ok(OutputFile {
            writer: Box::new(file),
            base_offset: actual_append,
            pos: 0,
        })
    } else {
        let file: Box<dyn WriteSeek> = if file_name == "-" {
            Box::new(std::io::Cursor::new(Vec::new()))
        } else {
            Box::new(std::fs::File::create(file_name)?)
        };
        Ok(OutputFile {
            writer: file,
            base_offset: 0,
            pos: 0,
        })
    }
}
